src/mesh/p4est_mesh.jl

```diff
@@ -0,0 +1,273 @@
+                            `faces[1]` describes the face onto which the face in negative x-direction
+                            of the unit hypercube is mapped. The face in positive x-direction of
+                            the unit hypercube will be mapped onto the face described by `faces[2]`.
+                            `faces[3:4]` describe the faces in positive and negative y-direction respectively
+                            (in 2D and 3D).
+                            `faces[5:6]` describe the faces in positive and negative z-direction respectively (in 3D).
+- `polydeg::Integer`: polynomial degree to used to store the geometry of the mesh.
+                      The mapping will be approximated by an interpolation polynomial
+                      of the specified degree for each tree.
+- `RealT::Type`: the type that should be used for coordinates.
+- `initial_refinement_level::Integer`: refine the mesh uniformly to this level before the simulation starts.
+- `periodicity`: either a `Bool` deciding if all of the boundaries are periodic or an `NTuple{NDIMS, Bool}` deciding for
+                 each dimension if the boundaries in this dimension are periodic.
+"""
+function P4estMesh(trees_per_dimension, faces::Tuple; polydeg, RealT=Float64, initial_refinement_level=0, periodicity=true)
+  validate_faces(faces)
+
+  # Use the transfinite mapping with the correct number of arguments
+  mapping = transfinite_mapping(faces)
+
+  return P4estMesh(trees_per_dimension, mapping; polydeg=polydeg, RealT=RealT,
+                   initial_refinement_level=initial_refinement_level,
+                   periodicity=periodicity, mapping_as_string=mapping_as_string)
+end
+
+
+"""
+    P4estMesh(trees_per_dimension, coordinates_min, coordinates_max;
+              polydeg, initial_refinement_level=0, periodicity=true)
+
+Create a P4estMesh that represents an uncurved structured mesh with a rectangular domain.
+
+# Arguments
+- `trees_per_dimension::NTuple{NDIMS, Int}`: the number of trees in each dimension.
+- `coordinates_min::NTuple{NDIMS, RealT}`: coordinate of the corner in the negative direction of each dimension.
+- `coordinates_max::NTuple{NDIMS, RealT}`: coordinate of the corner in the positive direction of each dimension.
+- `polydeg::Integer`: polynomial degree to used to store the geometry of the mesh.
+                      The mapping will be approximated by an interpolation polynomial
+                      of the specified degree for each tree.
+- `initial_refinement_level::Integer`: refine the mesh uniformly to this level before the simulation starts.
+- `periodicity`: either a `Bool` deciding if all of the boundaries are periodic or an `NTuple{NDIMS, Bool}` deciding for
+                 each dimension if the boundaries in this dimension are periodic.
+"""
+function P4estMesh(trees_per_dimension, coordinates_min, coordinates_max;
+                   polydeg, initial_refinement_level=0, periodicity=true)
```

Comment:
It might be a bit confusing and lead to hard to find bugs if people use vectors for the coordinates, since that will trigger another constructor. What do you think about making `coordinates_min, coordinates_max, faces, mapping` keyword arguments defaulting to `nothing`? Then, you could have one bigger constructor that would need to check that exactly one of the three options (coordinates, mapping, faces) is chosen, but the interface might be easier to remember, I think.

Reply:
This sounds like a very good idea.